### PR TITLE
Add Queue memory safety proofs.

### PR DESCRIPTION
**Pull request by markrtuttle**
Add Queue memory safety proofs originally written by Malte Mues and now rebased against the current master branch.

Review this request one commit at a time. This pull request changes a lot of files, but each commit introduces a single proof at a time. I have verified that the proofs work against the current master.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by markrtuttle** [comment]
The proofs in this pull request are works in progress.

We are committing them now because they are correct and have value, and because we want them to keep current with code changes.

But they make assumptions, and in the next pass we will do a better job of extracting the assumptions, stating the assumptions, and making the assumptions as weak as possible (to make the proofs as strong as possible).

File: tools/cbmc/proofs/Queue/QueueTakeMutexRecursive/QueueTakeMutexRecursive_harness.c [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+#include "queue.h"
+#include "tasksStubs.h"
+#include "queue_datastructure.h"
+#include "cbmc.h"
+
+/* prvUnlockQueue is going to decrement this value to 0 in the loop.
+   We need a bound for the loop. Using 5 has a reasonable performance resulting
+   in 4 unwinding iterations of the loop. The loop is mostly modifying a
+   data structure in task.c that is not in the scope of the proof. */
+#ifndef PRV_UNLOCK_UNWINDING_BOUND
```
Comment (karkhaz) [inline tools/cbmc/proofs/Queue/QueueTakeMutexRecursive/QueueTakeMutexRecursive_harness.c]: `PRV_UNLOCK_UNWINDING_BOUND` will always be defined, as it is set in `Makefile.json`. This block is therefore unnecessary.

File: tools/cbmc/proofs/Queue/QueueTakeMutexRecursive/QueueTakeMutexRecursive_harness.c [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	#define PRV_UNLOCK_UNWINDING_BOUND 5
+#endif
+
+/* Bound for the timeout cycles */
+#ifndef QueueSemaphoreTake_BOUND
```
Comment (karkhaz) [inline tools/cbmc/proofs/Queue/QueueTakeMutexRecursive/QueueTakeMutexRecursive_harness.c]: I don't really understand the logic of these defines. Some of them are written in the C file and others in `Makefile.json`. Is there a good reason for this? If not, then I can clean it up as part of the new build system rebase, no need to go through all of these right now IMHO.

File: tools/cbmc/include/queue_init.h [hunk 3]
```diff
@@ -2,6 +2,14 @@
 #include "queue.h"
 #include "queue_datastructure.h"
 
+#ifndef CBMC_OBJECT_BITS
+#define CBMC_OBJECT_BITS 7
```
[the remark this answers is not in the document]
Comment (markrtuttle) [reply tools/cbmc/include/queue_init.h]: It is my plan to strip out all of this stuff once your new build system is in place.  I prefer the naming scheme I'm using here (this naming scheme was one of my proposals in my review of your build system).

File: tools/cbmc/patches/0005-remove-static-from-prvCopyDataToQueue.patch [hunk 4]
```diff
@@ -0,0 +1,22 @@
+diff --git a/freertos_kernel/queue.c b/freertos_kernel/queue.c
+index df2b9c8c..c2265c26 100644
+--- a/freertos_kernel/queue.c
++++ b/freertos_kernel/queue.c
+@@ -105,7 +105,7 @@ static BaseType_t prvIsQueueFull( const Queue_t *pxQueue ) PRIVILEGED_FUNCTION;
+  * Copies an item into the queue, either at the front of the queue or the
+  * back of the queue.
+  */
+-static BaseType_t prvCopyDataToQueue( Queue_t * const pxQueue, const void *pvItemToQueue, const BaseType_t xPosition ) PRIVILEGED_FUNCTION;
```
[the remark this answers is not in the document]
Comment (markrtuttle) [reply tools/cbmc/patches/0005-remove-static-from-prvCopyDataToQueue.patch]: That is the plan.

File: tools/cbmc/proofs/Queue/QueueGenericCreate/Configurations.json [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  "CBMC_OBJECT_BITS": "7",
+  "CBMC_OBJECT_MAX_SIZE": "\"((UINT32_MAX>>(CBMC_OBJECT_BITS+1))\"",
+
+  "CBMCFLAGS":
+  [
+      "--unwind 1",
+      "--signed-overflow-check",
+      "--pointer-overflow-check",
+      "--unsigned-overflow-check"
+
```
[the remark this answers is not in the document]
Comment (markrtuttle) [reply tools/cbmc/proofs/Queue/QueueGenericCreate/Configurations.json]: I'm trying to keep whitespace changes out of this pass.

File: tools/cbmc/proofs/Queue/QueueGenericSendFromISR/QueueGenericSendFromISR_harness.c [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+
+	if( xQueue ){
+		void *pvItemToQueue = pvPortMalloc(xQueue->uxItemSize);
+		BaseType_t *xHigherPriorityTaskWoken = pvPortMalloc(sizeof(BaseType_t));
+		BaseType_t xCopyPosition;
+		if(xQueue->uxItemSize == 0)
+		{
+			/* uxQueue->xQueueType is a pointer to the head of the queue storage area.
+			   If an item has a sice, this pointer must not be modified after init.
```
[the remark this answers is not in the document]
Comment (markrtuttle) [reply tools/cbmc/proofs/Queue/QueueGenericSendFromISR/QueueGenericSendFromISR_harness.c]: Fixed.

File: tools/cbmc/proofs/Queue/QueueGiveMutexRecursive/QueueGiveMutexRecursive_harness.c [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+void harness() {
+  uint8_t ucQueueType;
+  QueueHandle_t xMutex =
+      xQueueCreateMutex( ucQueueType);
+  if (xMutex) {
+    xMutex->uxQueueType = ucQueueType;
+    UBaseType_t uxCounter;
+    /* This assumption is explained in the queue.c file inside the method body
+       xQueueGiveMutexRecursive and guards against an underflow error. */
```
[the remark this answers is not in the document]
Comment (markrtuttle) [reply tools/cbmc/proofs/Queue/QueueGiveMutexRecursive/QueueGiveMutexRecursive_harness.c]: Not just here.  I think that part of our next pass should be that every assumption made in the proofs (except perhaps bounds required by the capacity of CBMC) should be checked with assertions in the code itself.

File: tools/cbmc/proofs/Queue/QueueSemaphoreTake/QueueSemaphoreTake_harness.c [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+#include "queue.h"
+#include "queue_init.h"
+#include "tasksStubs.h"
+#include "cbmc.h"
+
+/* prvUnlockQueue is going to decrement this value to 0 in the loop.
+    We need a bound for the loop. Using 5 has a reasonable performance resulting
+    in 4 unwinding iterations of the loop. The loop is mostly modifying a
+    data structure in task.c that is not in the scope of the proof. */
+#ifndef PRV_UNLOCK_UNWINDING_BOUND
```
[the remark this answers is not in the document]
Comment (markrtuttle) [reply tools/cbmc/proofs/Queue/QueueSemaphoreTake/QueueSemaphoreTake_harness.c]: Deleing, and adding comment to Makefile.json.

File: tools/cbmc/proofs/Queue/QueueTakeMutexRecursive/QueueTakeMutexRecursive_harness.c (hunk 1 above)
Comment (markrtuttle) [reply tools/cbmc/proofs/Queue/QueueTakeMutexRecursive/QueueTakeMutexRecursive_harness.c]: Deleting, and adding comments to Makefile.json

File: tools/cbmc/proofs/Queue/QueueTakeMutexRecursive/QueueTakeMutexRecursive_harness.c (hunk 2 above)
Comment (markrtuttle) [reply tools/cbmc/proofs/Queue/QueueTakeMutexRecursive/QueueTakeMutexRecursive_harness.c]: Deleting, and adding comments to Makefile.json.

File: tools/cbmc/proofs/Queue/QueueGenericReset/README.md [hunk 9]
```diff
@@ -0,0 +1,12 @@
+Assuming that the QueueHandel_t is not null and the assumptions made
```
Comment (dcgaws) [inline tools/cbmc/proofs/Queue/QueueGenericReset/README.md]: QueueHandle_t

File: tools/cbmc/proofs/Queue/QueueCreateCountingSemaphoreStatic/README.md [hunk 10]
```diff
@@ -0,0 +1,11 @@
+Assuming uxMaxCount > 0, uxInitialCount <= uxMaxCount and the reference
+to the storage area is not null,
+this harness proves the memory saftey of QueueCreateCountingSemphoreStatic.
```
Comment (dcgaws) [inline tools/cbmc/proofs/Queue/QueueCreateCountingSemaphoreStatic/README.md]: Semaphore

File: tools/cbmc/proofs/Queue/QueueGenericReset/README.md (hunk 9 above)
Comment (karkhaz) [reply tools/cbmc/proofs/Queue/QueueGenericReset/README.md]: done, thank you

File: tools/cbmc/proofs/Queue/QueueCreateCountingSemaphoreStatic/README.md (hunk 10 above)
Comment (karkhaz) [reply tools/cbmc/proofs/Queue/QueueCreateCountingSemaphoreStatic/README.md]: done

**Pull request merged**